tests/recovery/blobmanipulator/blobmanipulator.go

```diff
@@ -0,0 +1,269 @@
+//go:build darwin || (linux && amd64)
+// +build darwin linux,amd64
+
+// Package blobmanipulator provides the framework for snapshot fix testing.
+package blobmanipulator
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"log"
+	"math/rand"
+	"strconv"
+	"strings"
+
+	"github.com/kopia/kopia/repo/blob"
+	"github.com/kopia/kopia/snapshot"
+	"github.com/kopia/kopia/tests/robustness"
+	"github.com/kopia/kopia/tests/robustness/fiofilewriter"
+	"github.com/kopia/kopia/tests/robustness/snapmeta"
+	"github.com/kopia/kopia/tests/tools/fio"
+	"github.com/kopia/kopia/tests/tools/kopiarunner"
+)
+
+// BlobManipulator provides a way to run a kopia command.
+type BlobManipulator struct {
+	KopiaCommandRunner *kopiarunner.KopiaSnapshotter
+	DirCreater         *snapmeta.KopiaSnapshotter
+	fileWriter         *fiofilewriter.FileWriter
+
+	DataRepoPath string
+}
+
+// NewBlobManipulator instantiates a new BlobManipulator and returns its pointer.
+func NewBlobManipulator(baseDirPath string) (*BlobManipulator, error) {
+	runner, err := kopiarunner.NewKopiaSnapshotter(baseDirPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BlobManipulator{
+		KopiaCommandRunner: runner,
+	}, nil
+}
+
+// ConnectOrCreateRepo connects to an existing repository if possible or creates a new one.
+func (bm *BlobManipulator) ConnectOrCreateRepo(dataRepoPath string) error {
+	return bm.DirCreater.ConnectOrCreateRepo(dataRepoPath)
```

Comment:
`DirCreater` may be `nil` if call that sets it fails.  You must check at the point that `DirCreater` is set for errors.  Do not continue to this line of code if `DirCreater` is `nil`.

Reply:
done